Should link protobuf (prevent #3)

esetomo opened the pull request:
When "-lprotobuf" can not be added to LIBS, installation is canceled.
This prevents installation of broken binary.

esetomo committed:
Should link protobuf (prevent #3)

When "-lprotobuf" can not be added to LIBS, installation is canceled.
This prevents installation of broken binary.

akihikodaki commented on ext/cld3/extconf.rb:
```diff
@@ -17,7 +17,9 @@
 
 # Check pkg-config first to inform the library is missing if so.
-pkg_config("protobuf")
+unless pkg_config("protobuf") || have_library("protobuf")
```
`have_library` should be removed because it ignores `pkg-config` tells. `have_library` is not documented well, probably because it could be environment-dependent, but it actually just adds `-lprotobuf` and tries to compile. See
https://github.com/ruby/ruby/blob/7d4cdab8f8c7e1f3f7e5c2add712659057f643fc/lib/mkmf.rb#L927.
Reading `pkg-config` is more reliable than appending `-lprotobuf`, which could miss additional flags required for compiling and linking, and that is why `pkg-config` exists.

akihikodaki commented on ext/cld3/extconf.rb:
```diff
@@ -17,7 +17,9 @@
 # Check pkg-config first to inform the library is missing if so.
-pkg_config("protobuf")
+unless pkg_config("protobuf") || have_library("protobuf")
+  abort "Can't find protobuf\n"
```
`Can't find` sounds a bit odd. Could you change to `protobuf not found` `Failed to locate protobuf`, or something sort of, as many other CLI tools do?

akihikodaki reviewed (changes requested):
That is what we need. I will merge this change after changes I requested in this review are done.

esetomo committed:
Changed protobuf check according to comment

comment says 'Check pkg-config'.
It is not clear whether pkg-config can be used in all environments,
but other search methods are not target feature.

comment says 'inform the library is missing'.
inform is the target feature.

akihikodaki commented on ext/cld3/extconf.rb:
```diff
@@ -17,8 +17,14 @@
-unless pkg_config("protobuf") || have_library("protobuf")
-  abort "Can't find protobuf\n"
+def have_protobuf?
+  checking_for("protobuf by pkg-config") do
```
Unfortunately, `checking_for` is "internal use only." Please remove.
https://github.com/ruby/ruby/blob/7d4cdab8f8c7e1f3f7e5c2add712659057f643fc/lib/mkmf.rb#L940

esetomo commented:
too complex. rewrite.

akihikodaki commented:
I think it is OK even if it doesn't have a checking message because:
* RubyGems, which ignores `stdout`, is likely to be used for installation.
* mkmf emits `mkmf.log`, which is useful enough to investigate an issue which you encountered.
* Follow the rule of silence.
